### PR TITLE
feat: Cross Chain marketplace partIII

### PR DESCRIPTION
**Pull request by salaheldinsoliman**
# Description of change

Part III of the tutorial, finally

## Links to any relevant issues

#1663 

## Type of change

- Documentation Enhancement

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the [contribution guidelines](https://github.com/iota-wiki/iota-wiki/blob/main/.github/CONTRIBUTING.md) for this project
- [x] I have performed a self-review of my own changes
- [x] I have made sure that added/changed links still work
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+No extra configurations are needed for this part of the tutorial. You can use the same `hardhat.config.js` file from [Part I](https://wiki.iota.org/isc/tutorials/cross-chain-nft-marketplace-part-1/).
+
+```javascript reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/hardhat.config.js
+```
+
+
+## Contracts Architecture
+
+![alt text](../../../../../../static/img/tutorials/cross_chain_marketplace/Architecture-V3.png)
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:33]:
```suggestion

## Contracts Architecture

![Contract Archicteture](../../../../../../static/img/tutorials/cross_chain_marketplace/Architecture-V3.png)
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+Next to the marketplace contract, you will deploy additional contracts to handle cross-chain NFT transfers and messages. To put things into perspective, here is the flow of the cross-chain NFT buying process from a user on BNB Testnet to the marketplace on ShimmerEVM Testnet:
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:35]: I'm not sure `next` is the right word here.

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+ _nftAddress,
+ _tokenId
+ );
+
+        // Send the payload to the agent on Shimmer
+        _lzSend(_dstChainId, payload, refundAddress, address(0x0), adapterParams, msg.value);
+ }
+```
+
+2. `_nonblockingLzReceive`: This function is called by the LayerZero endpoint on the destination chain upon the receipt of a message. We override this function in the `CrossChainAgent` contract Testnet to call the `_notifyMarketPlace` function on the marketplace contract to complete the purchase.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:99]:
```suggestion
2. `_nonblockingLzReceive`: This function is called by the LayerZero endpoint on the destination chain upon the receipt of a message. You should override this function in the `CrossChainAgent` contract Testnet to call the `_notifyMarketPlace` function on the marketplace contract to complete the purchase.
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+```solidity reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/contracts/CrossChainAgent.sol
+```
+
+### NFTMarketPlaceV2.sol
+
+The `NFTMarketPlaceV2` contract is an updated version of the `NFTMarketPlace` contract that you deployed in [Part I](https://wiki.iota.org/isc/tutorials/cross-chain-nft-marketplace-part-1/). This version includes a new function `buyItemCrossChain` that is called by the `CrossChainAgent` contract on the ShimmerEVM Testnet to complete the purchase of an NFT.
+
+The function checks if the item is listed, and available, and if the user has enough balance to complete the purchase. It then transfers the NFT to the user by making a call to the NFT's proxy contract (The contract responsible for transferring the NFT cross-chain). The user's balance is updated, and the item's availability is set to false.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:140]:
```suggestion
The function checks if the item is listed and available and if the user has enough balance to complete the purchase. It then transfers the NFT to the user by calling the NFT's proxy contract (the contract responsible for transferring the NFT cross-chain). The user's balance is updated, and the item's availability is set to false.
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+2. Deploy an `ONFT721` contract on ShimmerEVM Testnet and BNB Testnet, and configure them to send and receive NFTs across chains.
+
+3. Mint an NFT (from the `ONFT` contract), allow the marketplace to spend it then list it on the marketplace.
+
+4. Deploy the `CrossChainAgent` contract to the BNB Testnet and the ShimmerEVM Testnet.
+
+5. Configure the `CrossChainAgent` contract on the BNB Testnet to send messages to the `CrossChainAgent` contract on the ShimmerEVM Testnet.
+
+6. Call the `buyCrossChain` function on the `CrossChainAgent` contract on the BNB Testnet to buy the NFT listed on the marketplace on the ShimmerEVM Testnet.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:190]: please add links to the respective headers

Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:190]: please add links to the respective headers

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+
+
+### Deploy NFTMarketPlaceV2
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:194]:
```suggestion

### Deploy NFTMarketPlaceV2

```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+```
+
+### Deploy ONFT721
+
+The next step is to deploy the `ONFT721` contract to the ShimmerEVM Testnet and the BNB Testnet. The `ONFT721` contract is an ERC721 contract that can send and receive NFTs across chains.
+
+```javascript reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/scripts/deploy_onft_shimmer.js
+```
+
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:214]:
```suggestion
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+You can run both scripts by executing the following commands:
+
+```bash
+npx hardhat run scripts/deploy_onft_shimmer.js --network shimmerevm-testnet
+npx hardhat run scripts/deploy_onft_bnb.js  --network bnbTestnet
+```
+
+### Set trusted remote of ONFT on BNB and Shimmer
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:226]:
```suggestion
### Set the Trusted Remote Addresses of ONFT on BNB and Shimmer
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+After deploying the ONFT contracts, you need to wire them up by caling `setTrustedRemoteAddress` on the `ONFT721` contract on the ShimmerEVM Testnet and BNB Testnet.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:228]: perhaps connect them works better than wire them up

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+
+```javascript reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/scripts/set_trusted_remote_shimmer.js
+```
+
+```javascript reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/scripts/set_trusted_remote_bnb.js
+```
+
+You can run the script by executing the following command:
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:238]:
```suggestion
You can run the script by executing the following commands:
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+
+```bash
+npx hardhat run scripts/set_trusted_remote_bnb.js --network bnbTestnet
+npx hardhat run scripts/set_trusted_remote_shimmer.js --network shimmerevm-testnet
+```
+
+### Mint an NFT and list it on the marketplace
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:245]:
```suggestion
### Mint an NFT and List It on the Marketplace
```
https://capitalizemytitle.com/

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+
+After deployment and configuration of the `ONFT721` contracts, you can mint an NFT, allow the marketplace to spend it, and list it on the marketplace.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:247]:
```suggestion
After you deploy and configure the `ONFT721` contracts, you can mint an NFT, allow the marketplace to spend it, and list it on the marketplace using the following scripts:
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+```
+
+You can run the scripts by executing the following commands:
+
+```bash
+npx hardhat run scripts/mint_nft_shimmer.js --network shimmerevm-testnet
+npx hardhat run scripts/approve_myERC721_for_marketplace.js --network shimmerevm-testnet
+```
+
+### Deploy CrossChainAgent
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:264]:
```suggestion
### Deploy the `CrossChainAgent` Contract
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+Now that we have an NFT listed on the marketplace, we need to deploy the `CrossChainAgent` contract to the BNB Testnet and the ShimmerEVM Testnet.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:266]:
```suggestion
Once you have an NFT listed on the marketplace, you need to deploy the `CrossChainAgent` contract to the BNB Testnet and the ShimmerEVM Testnet.
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+```
+
+You can run the scripts by executing the following commands:
+
+```bash
+npx hardhat run scripts/deploy_messanger_shimmer.js --network shimmerevm-testnet
+npx hardhat run scripts/deploy_messanger_bnb.js --network bnbTestnet
+```
+
+### Configure CrossChainAgent
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:284]:
```suggestion
### Configure the `CrossChainAgent` Contracts
```

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+
+
+:::note
+For the OFTs, we are using the `MyOFT` contract, which is a simple ERC20 token contract. You can deploy this contract to the BNB Testnet and the ShimmerEVM Testnet using the same process as the `ONFT721` contract.
+:::
+
+
+
+## Conclusion
+In the third part of the tutorial, we have automated the buy flow of the cross-chain NFT marketplace. We have deployed the necessary contracts on the BNB Testnet and the ShimmerEVM Testnet and configured them to send and receive NFTs across chains.
```
Comment (lucas-tortora) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:324]:
```suggestion

In the third part of the tutorial, you have automated the buy flow of the cross-chain NFT marketplace. You deployed the necessary contracts on the BNB Testnet and the ShimmerEVM Testnet and configured them to send and receive NFTs across chains. 
```

**Review by lucas-tortora — APPROVED** [review]

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+
+This final part will evolve the marketplace's architecture to handle cross-chain NFT transfers. You will deploy the necessary contracts on the BNB Testnet and the ShimmerEVM Testnet and configure them to send and receive NFTs across chains. You will also create scripts to automate sending and receiving NFTs across chains.
+
+## Prerequisites
+
+- [Node.js](https://nodejs.org) >=  v18.0
+- [Hardhat](https://hardhat.org) >= v2.0.0
+- [npx](https://www.npmjs.com/package/npx)  >= v7.1.0.
+- The code for this part of the tutorial is based on [Part I](https://wiki.iota.org/isc/tutorials/cross-chain-nft-marketplace-part-1/) and [Part II](https://wiki.iota.org/isc/tutorials/cross-chain-nft-marketplace-part-2/) of the tutorial series.
+
```
Comment (vivekjain23) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:17]: Is it required to have prerequisites in Part 2 and Part 3 if already mentioned in Part 1?

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+```
+
+The full contract code is as follows:
+
+```solidity reference
+https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/contracts/NFTMarketPlaceV2.sol
+```
+
+
+## Scripts
```
Comment (vivekjain23) [inline docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:172]: In the scripts section, we could also implement a command to verify the deployed contracts, wdyt?

**Review by vivekjain23 — CHANGES_REQUESTED** [review]

File: docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md (hunk 18 above)
Comment (salaheldinsoliman) [reply docs/build/isc/v1.3/docs/tutorials/cross-chain-nft-marketplace-part-3.md:172]: That would be a good addition yes, but I think it is out of scope for this tutorial. 
We would need the APIs for ShimmerEVM and BNBTestnet blockexplorers, and add some extra config to `hardhat.config.js`.
What would be more suitable is adding this part later on, WDYT?

**Comment by lucas-tortora** [comment]
@salaheldinsoliman , are there any blockers on this? It'd be nice to publish the last part of the series :)

**Comment by salaheldinsoliman** [comment]
@vivekjain23 Can you please approve this If you don't have further comments? thanks!

**Review by vivekjain23 — APPROVED** [review]

**Pull request merged**